pallets/asset-depository/src/lib.rs

```diff
@@ -0,0 +1,175 @@
+        pub(crate) fn set_available_balance(
+            asset_id: &T::AssetId,
+            who: &AccountIdFor<T>,
+            amount: T::Balance,
+        ) {
+            Accounts::<T>::mutate(who, asset_id, |account| {
+                account.available = amount;
+            });
+        }
+    }
+
+    impl<T: Config> MultiAssetDepository<T::AssetId, AccountIdFor<T>, T::Balance> for Pallet<T> {
+        /// The total amount of the given asset currently held
+        fn aggregated_balance(asset_id: &T::AssetId) -> T::Balance {
+            TotalBalance::<T>::get(asset_id)
+        }
+
+        /// The total balance of an asset of a user
+        fn total_balance(asset_id: &T::AssetId, who: &AccountIdFor<T>) -> T::Balance {
+            Accounts::<T>::get(who, asset_id).total_balance()
+        }
+
+        /// The current available balance of an asset of a user
+        fn available_balance(asset_id: &T::AssetId, who: &AccountIdFor<T>) -> T::Balance {
+            Accounts::<T>::get(who, asset_id).available
+        }
+
+        /// Deposit the `amount` of the given asset into the available balance of the given account `who`.
+        fn deposit(
+            asset_id: &T::AssetId,
+            who: &AccountIdFor<T>,
+            amount: T::Balance,
+        ) -> DispatchResult {
+            if amount.is_zero() {
+                return Ok(());
+            }
+
+            TotalBalance::<T>::try_mutate(asset_id, |total| -> DispatchResult {
+                *total = total
+                    .checked_add(&amount)
+                    .ok_or(Error::<T>::TotalBalanceOverflow)?;
+
+                // SAFETY: this can't overflow because the balance for an account is
+                //  at most equal to the total balance and adding to it already succeeded
+                Self::set_available_balance(
+                    &asset_id,
+                    who,
+                    Self::available_balance(&asset_id, who) + amount,
+                );
+
+                Ok(())
+            })
+        }
+
+        /// Withdraw the `amount` of the given asset from the available balance of the given account `who`.
+        fn withdraw(
+            asset_id: &T::AssetId,
+            who: &AccountIdFor<T>,
+            amount: T::Balance,
+        ) -> DispatchResult {
+            if amount.is_zero() {
+                return Ok(());
+            }
+
+            Accounts::<T>::try_mutate(who, &asset_id, |balance| -> DispatchResult {
+                balance.available = balance
+                    .available
+                    .checked_sub(&amount)
+                    .ok_or(Error::<T>::NotEnoughBalance)?;
+                Ok(())
+            })?;
+
+            // SAFETY: this can't underflow because the total balance for an asset is at least
+            //  equal to the total balance in the given account and subtracting already succeeded
+            TotalBalance::<T>::mutate(asset_id, |total| *total -= amount);
```

Comment:
https://substrate.dev/rustdocs/v3.0.0/sp_arithmetic/traits/trait.Saturating.html#tymethod.saturating_sub

Reply:
I decided against saturating operations because values higher than the maximum are simply capped and that is not our intention.